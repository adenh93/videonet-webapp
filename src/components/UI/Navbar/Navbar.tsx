import React from "react";
import { createBrowserHistory } from "history";
import {
  Nav,
  Brand,
  NavItemList,
  NavItem,
  ProfileImage,
  NavRightContainer,
} from "./Styles";
import { NavRoute } from "../../../types/interfaces";

export interface NavbarProps {
  brandImg?: string;
  routes?: NavRoute[];
  profileImg?: string;
}

const Navbar: React.FC<NavbarProps> = ({ brandImg, routes, profileImg }) => {
  const history = createBrowserHistory();

  const onClickBrand = () => {
    history.push("/");
  };

  const onClickNavItem = (route: string) => {
    history.push(route);
  };

  const navItems = routes && (
    <NavItemList>
      {routes.map(({ display, route }) => (
        <NavItem
          key={route}
          active={history.location.pathname === route}
          onClick={() => onClickNavItem(route)}
        >
          {display}
        </NavItem>
      ))}
    </NavItemList>
  );

  return (
    <Nav>
      {brandImg && <Brand src={brandImg} onClick={onClickBrand} />}
      <NavRightContainer>
        {navItems}
        <ProfileImage src={profileImg} />
      </NavRightContainer>
    </Nav>
  );
};

export default Navbar;
