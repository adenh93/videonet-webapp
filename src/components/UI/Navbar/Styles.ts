import styled from "styled-components";

export const Nav = styled.nav`
  position: absolute;
  display: inline-block;
  width: calc(100% - ${({ theme }) => theme.margins.gutter * 2 + 2}rem);
  z-index: 3;
  top: 2rem;
  left: 2rem;
`;

export const Brand = styled.img`
  cursor: pointer;
  width: 21.5rem;
  height: 6rem;
`;

export const NavRightContainer = styled.div`
  float: right;
`;

export const NavItemList = styled.ul`
  display: inline-block;
  vertical-align: top;
  transform: translateY(60%);
  list-style: none;
`;

interface NavItemProps {
  active: boolean;
}

export const NavItem = styled.li<NavItemProps>`
  font-size: ${({ theme }) => theme.fontSize.navItem};
  cursor: pointer;
  letter-spacing: 0.03em;
  display: inline;
  padding: 0.5rem 1.3rem;
  border-bottom: ${({ active, theme }) =>
    active ? `2px solid ${theme.palette.text.main}` : "none"};
  margin-right: 6rem;
  &:hover {
    border-bottom: 2px solid ${({ theme }) => theme.palette.text.main};
  }
`;

export const ProfileImage = styled.img`
  border-radius: 20rem;
  width: 6rem;
`;
