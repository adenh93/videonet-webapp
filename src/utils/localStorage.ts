import { FilmDetails, SearchResult } from "../graphql/types";

type LocalStorageCollection = "WATCHED_ITEMS" | "LIKED_ITEMS";

export const setItem = <T>(type: LocalStorageCollection, item: T): void =>
  localStorage.setItem(type, JSON.stringify(item));

export const getItems = <R>(type: LocalStorageCollection): R[] => {
  const items = localStorage.getItem(type);
  if (!items) return [];
  return JSON.parse(items) as R[];
};

export const getWatchList = (): SearchResult[] =>
  getItems<SearchResult>("WATCHED_ITEMS");

export const isInWatchList = (movieId: number): boolean =>
  getWatchList().find((m) => m.id === movieId) !== undefined;

export const addToWatchList = (movie: FilmDetails): void => {
  const movies = getWatchList();
  if (!movies.find((m) => m.id === movie.id))
    setItem("WATCHED_ITEMS", [...movies, movie]);
};

export const removeFromWatchList = (movieId: number): void =>
  setItem(
    "WATCHED_ITEMS",
    getWatchList().filter((m) => m.id !== movieId)
  );

export const getLikedMovies = (): number[] => getItems<number>("LIKED_ITEMS");

export const isLiked = (movieId: number): boolean =>
  getLikedMovies().find((item) => item === movieId) !== undefined;

export const likeMovie = (movieId: number): void => {
  const likedItems = getLikedMovies();
  if (!likedItems.find((item) => item === movieId))
    setItem("LIKED_ITEMS", [...likedItems, movieId]);
};

export const unlikeMovie = (movieId: number): void => {
  setItem(
    "LIKED_ITEMS",
    getLikedMovies().filter((item) => item !== movieId)
  );
};

export const toggleWatched = (movie: FilmDetails): boolean => {
  const watched = isInWatchList(movie.id);
  watched ? removeFromWatchList(movie.id) : addToWatchList(movie);
  return !watched;
};

export const toggleLiked = (movieId: number): boolean => {
  const liked = isLiked(movieId);
  liked ? unlikeMovie(movieId) : likeMovie(movieId);
  return !liked;
};
