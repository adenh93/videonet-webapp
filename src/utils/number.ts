import numeral from "numeral";

export const formatNumber = (n: Number): String => numeral(n).format("0,0");
