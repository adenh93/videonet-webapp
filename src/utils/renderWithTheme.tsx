import React, { ReactElement } from "react";
import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { BrowserRouter as Router } from "react-router-dom";
import theme from "../theme";

export const renderWithTheme = (component: ReactElement) =>
  renderer.create(
    <ThemeProvider theme={theme}>
      <Router>{component}</Router>
    </ThemeProvider>
  );

export const rtlWithTheme = (component: ReactElement) =>
  render(
    <ThemeProvider theme={theme}>
      <Router>{component}</Router>
    </ThemeProvider>
  );
