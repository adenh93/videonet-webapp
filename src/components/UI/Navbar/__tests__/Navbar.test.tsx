import React from "react";
import { renderWithTheme } from "../../../../utils/renderWithTheme";
import Navbar from "../Navbar";
import "jest-styled-components";

describe("Navbar Typography component", () => {
  it("renders correctly", () => {
    const tree = renderWithTheme(<Navbar>Test</Navbar>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
