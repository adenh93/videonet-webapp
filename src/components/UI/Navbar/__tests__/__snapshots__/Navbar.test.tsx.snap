// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Typography component renders correctly 1`] = `
.c0 {
  position: absolute;
  display: inline-block;
  width: calc(100% - 4rem);
  z-index: 3;
  top: 2rem;
  left: 2rem;
}

.c1 {
  float: right;
}

.c2 {
  border-radius: 20rem;
  width: 6rem;
}

<nav
  className="c0"
>
  <div
    className="c1"
  >
    <img
      className="c2"
    />
  </div>
</nav>
`;
